## Part 2B: Optimisation and Machine Learning in Finance – Software - Solution

## Linear Regression

    In all the models used below we tried to classify the data into 16 categories and predicted whether the firm is a investment grade or not along with its accuracy in prediction. The datset consists of 1700 observations of 26 financial and accounting metrics for a set of firms in several different industries. Here we will first implement linear regression model to predict the ratings. The steps followed here are importing of libraries and loading of data sets. We then preprocess the data and then split the data into training and testing sets. We then train the model on the training data sets and test it on testing datasets to evaluate its performance. Use the trained model for new predictions.

In [1]:
# Importing necessary libraries

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge,Lasso, LinearRegression
from sklearn.metrics import accuracy_score, precision_score

In [2]:
# Loading of datasets
df = pd.read_csv('MLF_GP1_CreditScore.csv')
df.head()

Sales/Revenues  Gross Margin    EBITDA  EBITDA Margin  \
0       -0.005496      0.030763  0.018885       0.024515   
1       -0.005496      0.030763  0.088716       0.094733   
2       -0.007045      0.023159  0.088716       0.096440   
3       -0.009396      0.028400  0.088716       0.099046   
4       -0.009009      0.027714  0.088716       0.098611   

   Net Income Before Extras  Total Debt  Net Debt   LT Debt   ST Debt  \
0                  0.146849   -0.029710 -0.019296 -0.042648  0.049875   
1                  0.146849   -0.029710 -0.019296 -0.042648  0.049875   
2                  0.108590    0.039410  0.034268  0.009059  0.250371   
3                  0.146137    0.030071  0.036938 -0.016964  0.356994   
4                  0.123500    0.024224  0.034445 -0.034132  0.461894   

       Cash  ...  Interest Coverage  Total Liquidity  Current Liquidity  \
0 -0.133716  ...           0.136748         0.392143          -0.184887   
1 -0.133716  ...           0.214657         0.392143          -0.184887   
2  0.101315  ...           0.205290         0.483257          -0.017877   
3 -0.052606  ...           0.232991         0.996955          -0.122017   
4 -0.090869  ...           0.172906         1.711426          -0.161561   

   Current Liabilities  EPS Before Extras        PE       ROA       ROE  \
0             0.062781           0.148305  0.100409  0.163266  0.102521   
1             0.062781           0.148305 -0.089598  0.163266  0.102521   
2             0.121357           0.110656 -0.045142  0.105711  0.103378   
3             0.079051           0.151639 -0.008231  0.162421  0.132295   
4             0.084319           0.130435  0.015528  0.156427  0.225144   

   InvGrd  Rating  
0       1      A1  
1       1      A1  
2       1      A1  
3       1      A1  
4       1      A1  

[5 rows x 28 columns]

In [3]:
df.describe()

Sales/Revenues  Gross Margin       EBITDA  EBITDA Margin  \
count     1700.000000   1700.000000  1700.000000    1700.000000   
mean         0.050378      0.026007     0.068718       0.021074   
std          0.161910      0.273768     0.237365       0.189025   
min         -0.661715     -0.794722    -0.782254      -0.805153   
25%         -0.005693     -0.020028    -0.022640      -0.042771   
50%          0.034000      0.003403     0.049482       0.011134   
75%          0.083004      0.025595     0.124533       0.060566   
max          2.277229      3.202713     3.542425       4.141182   

       Net Income Before Extras   Total Debt     Net Debt      LT Debt  \
count               1700.000000  1700.000000  1700.000000  1700.000000   
mean                   0.123026     0.822405    -0.419810     1.255168   
std                   14.475689    13.317075    28.385702    16.224453   
min                 -289.000000    -0.903014  -493.305578    -0.921515   
25%                   -0.158478    -0.076316    -0.120725    -0.094767   
50%                    0.056627     0.005886    -0.003060    -0.002078   
75%                    0.222219     0.136449     0.160251     0.174735   
max                  478.280075   281.604237   865.194595   289.388178   

           ST Debt         Cash  ...          CFO  Interest Coverage  \
count  1700.000000  1700.000000  ...  1700.000000        1700.000000   
mean      3.142797     0.466620  ...    -0.189317           0.298785   
std      51.986550     1.859494  ...     5.668669           5.265291   
min      -0.997692    -0.990982  ...  -161.609425          -0.991976   
25%      -0.337959    -0.195117  ...    -0.115159          -0.096996   
50%       0.043092     0.075820  ...     0.046983           0.043216   
75%       0.649475     0.483113  ...     0.216432           0.177340   
max    2038.000000    36.980037  ...    13.005788         182.131887   

       Total Liquidity  Current Liquidity  Current Liabilities  \
count      1700.000000        1700.000000          1700.000000   
mean         -0.855714           0.436002             0.072802   
std          22.926862           1.904282             0.266471   
min        -502.000000          -0.994141            -0.684678   
25%          -0.857013          -0.227327            -0.072734   
50%          -0.229098           0.040446             0.041785   
75%           0.512778           0.416067             0.161215   
max         280.138728          34.372455             4.194381   

       EPS Before Extras           PE          ROA          ROE       InvGrd  
count        1700.000000  1700.000000  1700.000000  1700.000000  1700.000000  
mean            0.032196     0.497705     0.019394    -0.217604     0.757059  
std             6.151994    12.102502    14.594193    15.389000     0.428986  
min           -96.250000   -59.795133  -305.462167  -373.837267     0.000000  
25%            -0.152894    -0.293521    -0.208483    -0.233955     1.000000  
50%             0.066027    -0.040405    -0.009403    -0.020392     1.000000  
75%             0.236046     0.168897     0.156136     0.201596     1.000000  
max           187.000000   381.243282   474.847172   343.145356     1.000000  

[8 rows x 27 columns]

    The method used in spliting of training and testing sets is in the ratio of 80:20, which means 80% of the training data sets are used training and 20% of the data is used for testing. In this function the test size is meassured between 0 to 1, here 0.2 means 20% of the data sets. The random state controls the shuffling applied to the data before applying the split.

In [4]:
# Split the dataset into training and testing sets
x = df.iloc[:, :-2]  # select all columns except the last 2
y1 = df.iloc[:, -2]  # Select last but 2nd column
#y1 = df["InvGrd"]
#y2 = df.iloc[:, -1]  # Selects the last column
y2 = df["Rating"]
# Split arrays or matrices into random train and test subsets
x_train, x_test, y1_train, y1_test, y2_train, y2_test = train_test_split(x, y1, y2, test_size=0.2, 
                                                                         random_state=97, shuffle=False)

    We train the this model to solve regression problems where the loss function is identified by the least square function and regularization is given by the params. Alpha controls the strength of regularization. Decreasing aplha leads to lower variance and low bias and increasing alpha leads to high variance and high bias. Here we choose low alpha value. Once the reguarization is selected we then fit the regression model into it where we train the model and then we predict the value on test data sets. On analysis the accuracy was found to be 90%.

In [5]:
# Train Ridge model
ridge = Ridge(alpha = 0.01, fit_intercept=True,
    normalize=False,
    copy_X=True,
    max_iter=None,
    tol=0.001,
    solver='auto',
    random_state=None)  # Regression model
ridge.fit(x_train, y1_train, sample_weight=None)  # Regression fit
y1_predict_ridge = ridge.predict(x_test) # Predicting using the test data set

In [7]:
# Evaluate Ridge model
y1_predict_ridge_binary = [1 if x >= 0.5 else 0 for x in y1_predict_ridge]
# for x in y1_predict_ridge:
#     if x >= 0.5:
#         1
#     else:
#         0
ridge_accuracy = accuracy_score(y1_test, y1_predict_ridge_binary)
print("Ridge Accuracy:", round(ridge_accuracy,2))

Ridge Accuracy: 0.9


    We train the this model to solve regression problems where the loss function is identified by the least square function and regularization is given by the params. Alpha controls the strength of regularization. Decreasing aplha leads to lower variance and low bias and increasing alpha leads to high variance and high bias. Here we choose low alpha value once again. Once the reguarization is selected we then fit the regression model into it where we train the model and then we predict the value on test data sets. On analysis the accuracy was found to be 96%.

In [8]:
# Train Lasso model
lasso = Lasso(alpha=0.01, fit_intercept=True,
    normalize=False,
    precompute=False,
    copy_X=True,
    max_iter=1000,
    tol=0.0001,
    warm_start=False,
    positive=False,
    random_state=None,
    selection='cyclic') # Regression model
lasso.fit(x_train, y1_train) # Regression fit
y1_predict_lasso = lasso.predict(x_test)  # Predicting using the test data set

In [10]:
# Evaluate Lasso model
y1_predict_lasso_binary = [1 if x >= 0.5 else 0 for x in y1_predict_lasso]
# for x in y1_predict_lasso:
#     if x >= 0.5:
#         1
#     else:
#         0
lasso_accuracy = accuracy_score(y1_test, y1_predict_lasso_binary)
print("Lasso Accuracy:", round(lasso_accuracy,2))

Lasso Accuracy: 0.96


    In summary to prevent any overfitting inside the model training we used lasso and ridge. On comparing the accuracy between the two regressions lasso performs better with 96% accuracy. 

########################################################################################################################

## Logistic regression

    In this regression model used below we tried to classify the data into 16 categories and predicted whether the firm is a investment grade or not along with its accuracy in prediction. We work with the same datset which consists of 1700 observations of 26 financial and accounting metrics for a set of firms in several different industries. The steps required here for logistic regression model are importing of libraries and loading of data sets. We then preprocess the data and then split the data into training and testing sets. We then train the model on the training data sets and test it on testing datasets to evaluate its performance. On top of this we conduct l1 and l2 regularization. Use the trained model for new predictions. The Splitting of the datasets into training and testing sets are in the ratio of 80:20. 

In [112]:
# Importing necessary libraries

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score

# Loading of data
data = pd.read_csv('MLF_GP1_CreditScore.csv')
data.head()

Sales/Revenues  Gross Margin    EBITDA  EBITDA Margin  \
0       -0.005496      0.030763  0.018885       0.024515   
1       -0.005496      0.030763  0.088716       0.094733   
2       -0.007045      0.023159  0.088716       0.096440   
3       -0.009396      0.028400  0.088716       0.099046   
4       -0.009009      0.027714  0.088716       0.098611   

   Net Income Before Extras  Total Debt  Net Debt   LT Debt   ST Debt  \
0                  0.146849   -0.029710 -0.019296 -0.042648  0.049875   
1                  0.146849   -0.029710 -0.019296 -0.042648  0.049875   
2                  0.108590    0.039410  0.034268  0.009059  0.250371   
3                  0.146137    0.030071  0.036938 -0.016964  0.356994   
4                  0.123500    0.024224  0.034445 -0.034132  0.461894   

       Cash  ...  Interest Coverage  Total Liquidity  Current Liquidity  \
0 -0.133716  ...           0.136748         0.392143          -0.184887   
1 -0.133716  ...           0.214657         0.392143          -0.184887   
2  0.101315  ...           0.205290         0.483257          -0.017877   
3 -0.052606  ...           0.232991         0.996955          -0.122017   
4 -0.090869  ...           0.172906         1.711426          -0.161561   

   Current Liabilities  EPS Before Extras        PE       ROA       ROE  \
0             0.062781           0.148305  0.100409  0.163266  0.102521   
1             0.062781           0.148305 -0.089598  0.163266  0.102521   
2             0.121357           0.110656 -0.045142  0.105711  0.103378   
3             0.079051           0.151639 -0.008231  0.162421  0.132295   
4             0.084319           0.130435  0.015528  0.156427  0.225144   

   InvGrd  Rating  
0       1      A1  
1       1      A1  
2       1      A1  
3       1      A1  
4       1      A1  

[5 rows x 28 columns]

In [113]:
data.describe()

Sales/Revenues  Gross Margin       EBITDA  EBITDA Margin  \
count     1700.000000   1700.000000  1700.000000    1700.000000   
mean         0.050378      0.026007     0.068718       0.021074   
std          0.161910      0.273768     0.237365       0.189025   
min         -0.661715     -0.794722    -0.782254      -0.805153   
25%         -0.005693     -0.020028    -0.022640      -0.042771   
50%          0.034000      0.003403     0.049482       0.011134   
75%          0.083004      0.025595     0.124533       0.060566   
max          2.277229      3.202713     3.542425       4.141182   

       Net Income Before Extras   Total Debt     Net Debt      LT Debt  \
count               1700.000000  1700.000000  1700.000000  1700.000000   
mean                   0.123026     0.822405    -0.419810     1.255168   
std                   14.475689    13.317075    28.385702    16.224453   
min                 -289.000000    -0.903014  -493.305578    -0.921515   
25%                   -0.158478    -0.076316    -0.120725    -0.094767   
50%                    0.056627     0.005886    -0.003060    -0.002078   
75%                    0.222219     0.136449     0.160251     0.174735   
max                  478.280075   281.604237   865.194595   289.388178   

           ST Debt         Cash  ...          CFO  Interest Coverage  \
count  1700.000000  1700.000000  ...  1700.000000        1700.000000   
mean      3.142797     0.466620  ...    -0.189317           0.298785   
std      51.986550     1.859494  ...     5.668669           5.265291   
min      -0.997692    -0.990982  ...  -161.609425          -0.991976   
25%      -0.337959    -0.195117  ...    -0.115159          -0.096996   
50%       0.043092     0.075820  ...     0.046983           0.043216   
75%       0.649475     0.483113  ...     0.216432           0.177340   
max    2038.000000    36.980037  ...    13.005788         182.131887   

       Total Liquidity  Current Liquidity  Current Liabilities  \
count      1700.000000        1700.000000          1700.000000   
mean         -0.855714           0.436002             0.072802   
std          22.926862           1.904282             0.266471   
min        -502.000000          -0.994141            -0.684678   
25%          -0.857013          -0.227327            -0.072734   
50%          -0.229098           0.040446             0.041785   
75%           0.512778           0.416067             0.161215   
max         280.138728          34.372455             4.194381   

       EPS Before Extras           PE          ROA          ROE       InvGrd  
count        1700.000000  1700.000000  1700.000000  1700.000000  1700.000000  
mean            0.032196     0.497705     0.019394    -0.217604     0.757059  
std             6.151994    12.102502    14.594193    15.389000     0.428986  
min           -96.250000   -59.795133  -305.462167  -373.837267     0.000000  
25%            -0.152894    -0.293521    -0.208483    -0.233955     1.000000  
50%             0.066027    -0.040405    -0.009403    -0.020392     1.000000  
75%             0.236046     0.168897     0.156136     0.201596     1.000000  
max           187.000000   381.243282   474.847172   343.145356     1.000000  

[8 rows x 27 columns]

In [114]:
# Split the data into training and testing sets
train_data, test_data = train_test_split(data, test_size=0.2, random_state=42)

    We split the data into training and testing sets, we then do data preprocessing where 80% of the training data sets are used for training and 20% of the data is used for testing. We then fit a logistic regression model with L1 regularization to the training data. We used liblinear solver and penalty l1 to predict the accuracy of the model. Upon prediction the acciracy of the model was found to be 76.47% 

In [115]:
# Split the dataset into training and testing sets
x_train = train_data.iloc[:, :-2]  # select all columns except the last 2
y_train = train_data.iloc[:, -2]   # Select last but 2nd column

x_test = test_data.iloc[:, :-2]   # select all columns except the last 2
y_test = test_data.iloc[:, -2]    # Select last but 2nd column

In [116]:
# Data preprocessing

# Fit a logistic regression model with L1 regularization to the training data
Ridge = LogisticRegression(penalty='l1',solver='liblinear', max_iter=100,
    multi_class='auto',
    verbose=0,
    warm_start=False,
    n_jobs=None,
    l1_ratio=None)   # Logistic regression classifier
Ridge.fit(x_train, y_train)  # Regression fit

LogisticRegression(penalty='l1', solver='liblinear')

In [117]:
# Model prediction
p_y = Ridge.predict(x_test)

In [118]:
accuracy = accuracy_score(y_test, p_y)         # L1 accuracy
print("Ridge Accuracy:", round(accuracy,4))

Ridge Accuracy: 0.7647


In [119]:
# Data pre processing

# Fit a logistic regression model with L2 regularization to the training data
Lasso = LogisticRegression(penalty='l2',solver='newton-cg',
                           max_iter=100,
    multi_class='auto',
    verbose=0,
    warm_start=False,
    n_jobs=None,
    l1_ratio=None)    # Logistic regression classifier
Lasso.fit(x_train, y_train)  # Regression fit

LogisticRegression(solver='newton-cg')

In [120]:
# Lasso prediction
p_y = Lasso.predict(x_test)

In [121]:
accuracy = accuracy_score(y_test, p_y)      # check the accuracy score
print("Lasso Accuracy:", round(accuracy,4))

Lasso Accuracy: 0.7676


    Upon splitting up of data and conducting l1 regularization technique to find the best possible fit, here we find the accuracy of the Lasso regularisation where we use a different solver known as newton-cg to improve the accuracy of the model. Upon fitting of data into the training model we then predict that the accuracy of the model improved marginally. By this we can conclude that both the regularization predict that a firm is of a investment grade. The accuracy in predicting such a grade is 76.6% in both Rigid and lasso regularization technique.

##########################################################################################################################

## Neural Network

    Neural network learning based aprroaches are to classify firms rating to any one of the categories and check if it is an investment grade. The datset consists of 1700 observations of 26 financial and accounting metrics for a set of firms in several different industries. The steps followed here are importing of libraries and loading of data sets. Here we will preprocess the data by removing NA's or converting categorical to numeric.  We then split the data into training and testing sets. We then train the model on the training data sets and test it on testing datasets to evaluate its performance. Use the trained neural network model for new predictions.

In [122]:
# Importing necessary libraries

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from keras.models import Sequential
from keras.layers import Dense

In [123]:
# Loading of dataset
data = pd.read_csv("MLF_GP1_CreditScore.csv")
data.head()

Sales/Revenues  Gross Margin    EBITDA  EBITDA Margin  \
0       -0.005496      0.030763  0.018885       0.024515   
1       -0.005496      0.030763  0.088716       0.094733   
2       -0.007045      0.023159  0.088716       0.096440   
3       -0.009396      0.028400  0.088716       0.099046   
4       -0.009009      0.027714  0.088716       0.098611   

   Net Income Before Extras  Total Debt  Net Debt   LT Debt   ST Debt  \
0                  0.146849   -0.029710 -0.019296 -0.042648  0.049875   
1                  0.146849   -0.029710 -0.019296 -0.042648  0.049875   
2                  0.108590    0.039410  0.034268  0.009059  0.250371   
3                  0.146137    0.030071  0.036938 -0.016964  0.356994   
4                  0.123500    0.024224  0.034445 -0.034132  0.461894   

       Cash  ...  Interest Coverage  Total Liquidity  Current Liquidity  \
0 -0.133716  ...           0.136748         0.392143          -0.184887   
1 -0.133716  ...           0.214657         0.392143          -0.184887   
2  0.101315  ...           0.205290         0.483257          -0.017877   
3 -0.052606  ...           0.232991         0.996955          -0.122017   
4 -0.090869  ...           0.172906         1.711426          -0.161561   

   Current Liabilities  EPS Before Extras        PE       ROA       ROE  \
0             0.062781           0.148305  0.100409  0.163266  0.102521   
1             0.062781           0.148305 -0.089598  0.163266  0.102521   
2             0.121357           0.110656 -0.045142  0.105711  0.103378   
3             0.079051           0.151639 -0.008231  0.162421  0.132295   
4             0.084319           0.130435  0.015528  0.156427  0.225144   

   InvGrd  Rating  
0       1      A1  
1       1      A1  
2       1      A1  
3       1      A1  
4       1      A1  

[5 rows x 28 columns]

In [124]:
data.describe()

Sales/Revenues  Gross Margin       EBITDA  EBITDA Margin  \
count     1700.000000   1700.000000  1700.000000    1700.000000   
mean         0.050378      0.026007     0.068718       0.021074   
std          0.161910      0.273768     0.237365       0.189025   
min         -0.661715     -0.794722    -0.782254      -0.805153   
25%         -0.005693     -0.020028    -0.022640      -0.042771   
50%          0.034000      0.003403     0.049482       0.011134   
75%          0.083004      0.025595     0.124533       0.060566   
max          2.277229      3.202713     3.542425       4.141182   

       Net Income Before Extras   Total Debt     Net Debt      LT Debt  \
count               1700.000000  1700.000000  1700.000000  1700.000000   
mean                   0.123026     0.822405    -0.419810     1.255168   
std                   14.475689    13.317075    28.385702    16.224453   
min                 -289.000000    -0.903014  -493.305578    -0.921515   
25%                   -0.158478    -0.076316    -0.120725    -0.094767   
50%                    0.056627     0.005886    -0.003060    -0.002078   
75%                    0.222219     0.136449     0.160251     0.174735   
max                  478.280075   281.604237   865.194595   289.388178   

           ST Debt         Cash  ...          CFO  Interest Coverage  \
count  1700.000000  1700.000000  ...  1700.000000        1700.000000   
mean      3.142797     0.466620  ...    -0.189317           0.298785   
std      51.986550     1.859494  ...     5.668669           5.265291   
min      -0.997692    -0.990982  ...  -161.609425          -0.991976   
25%      -0.337959    -0.195117  ...    -0.115159          -0.096996   
50%       0.043092     0.075820  ...     0.046983           0.043216   
75%       0.649475     0.483113  ...     0.216432           0.177340   
max    2038.000000    36.980037  ...    13.005788         182.131887   

       Total Liquidity  Current Liquidity  Current Liabilities  \
count      1700.000000        1700.000000          1700.000000   
mean         -0.855714           0.436002             0.072802   
std          22.926862           1.904282             0.266471   
min        -502.000000          -0.994141            -0.684678   
25%          -0.857013          -0.227327            -0.072734   
50%          -0.229098           0.040446             0.041785   
75%           0.512778           0.416067             0.161215   
max         280.138728          34.372455             4.194381   

       EPS Before Extras           PE          ROA          ROE       InvGrd  
count        1700.000000  1700.000000  1700.000000  1700.000000  1700.000000  
mean            0.032196     0.497705     0.019394    -0.217604     0.757059  
std             6.151994    12.102502    14.594193    15.389000     0.428986  
min           -96.250000   -59.795133  -305.462167  -373.837267     0.000000  
25%            -0.152894    -0.293521    -0.208483    -0.233955     1.000000  
50%             0.066027    -0.040405    -0.009403    -0.020392     1.000000  
75%             0.236046     0.168897     0.156136     0.201596     1.000000  
max           187.000000   381.243282   474.847172   343.145356     1.000000  

[8 rows x 27 columns]

In [125]:
# Converting factoral to numeric
e = LabelEncoder()
data["Rating"] = e.fit_transform(data["Rating"])

    We split the data into training and testing sets, we then do data preprocessing where 80% of the training data sets are used for training and 20% of the data is used for testing. 

In [126]:
# Spliting the data into training and testing
train_data, test_data = train_test_split(data, test_size=0.2, random_state=42, , shuffle=False)

SyntaxError: invalid syntax (<ipython-input-126-8c886d07b54b>, line 2)

In [ ]:
# Locating the variables 
x_train = train_data.iloc[:, :-2].values       # select all columns except the last 2
y_train = train_data.iloc[:, -2:].values[:, 0]  # Select last but 2nd column
x_test = test_data.iloc[:, :-2].values          # select all columns except the last 2
y_test = test_data.iloc[:, -2:].values[:, 0]    # Selects last 2 columns

In [ ]:
# Defining model architecture
m = Sequential() # defined from keras model 
m.add(Dense(units=64, activation="relu", input_dim=x_train.shape[1]))   # dense by keras layer
m.add(Dense(units=32, activation="relu"))                     # dense by keras layer
m.add(Dense(units=16, activation="softmax"))                   # dense by keras layer

In [127]:
# Model compilation
m.compile(optimizer="rmsprop", loss="sparse_categorical_crossentropy", metrics=["accuracy"], loss_weights=None,
          weighted_metrics=None, run_eagerly=None, steps_per_execution=None,
    jit_compile=None)

In [128]:
# Training of data set
m.fit(x_train, y_train, epochs=100, batch_size=64, validation_split=0.2)

Epoch 1/100
17/17 [==============================] - 1s 30ms/step - loss: 0.0749 - accuracy: 0.9853 - val_loss: 0.6191 - val_accuracy: 0.8382
Epoch 2/100
17/17 [==============================] - 0s 8ms/step - loss: 0.0477 - accuracy: 0.9871 - val_loss: 0.6403 - val_accuracy: 0.8382
Epoch 3/100
17/17 [==============================] - 0s 8ms/step - loss: 0.0504 - accuracy: 0.9871 - val_loss: 0.6989 - val_accuracy: 0.8419
Epoch 4/100
17/17 [==============================] - 0s 13ms/step - loss: 0.0538 - accuracy: 0.9853 - val_loss: 0.7043 - val_accuracy: 0.8456
Epoch 5/100
17/17 [==============================] - 0s 13ms/step - loss: 0.0524 - accuracy: 0.9881 - val_loss: 0.6519 - val_accuracy: 0.8529
Epoch 6/100
17/17 [==============================] - 0s 13ms/step - loss: 0.0559 - accuracy: 0.9871 - val_loss: 0.7686 - val_accuracy: 0.8493
Epoch 7/100
17/17 [==============================] - 0s 12ms/step - loss: 0.0608 - accuracy: 0.9871 - val_loss: 0.6607 - val_accuracy: 0.8382
Epoch 8/

17/17 [==============================] - 0s 7ms/step - loss: 0.0235 - accuracy: 0.9972 - val_loss: 0.7884 - val_accuracy: 0.8456
Epoch 59/100
17/17 [==============================] - 0s 7ms/step - loss: 0.0446 - accuracy: 0.9899 - val_loss: 0.7471 - val_accuracy: 0.8382
Epoch 60/100
17/17 [==============================] - 0s 7ms/step - loss: 0.0332 - accuracy: 0.9963 - val_loss: 0.7790 - val_accuracy: 0.8419
Epoch 61/100
17/17 [==============================] - 0s 7ms/step - loss: 0.0246 - accuracy: 0.9963 - val_loss: 0.7725 - val_accuracy: 0.8346
Epoch 62/100
17/17 [==============================] - 0s 8ms/step - loss: 0.0311 - accuracy: 0.9917 - val_loss: 0.7248 - val_accuracy: 0.8346
Epoch 63/100
17/17 [==============================] - 0s 7ms/step - loss: 0.0255 - accuracy: 0.9963 - val_loss: 0.8042 - val_accuracy: 0.8382
Epoch 64/100
17/17 [==============================] - 0s 7ms/step - loss: 0.0253 - accuracy: 0.9954 - val_loss: 0.7402 - val_accuracy: 0.8419
Epoch 65/100
17/17 

In [129]:
# Model testing
test_loss, test_acc = m.evaluate(x_test, y_test,
    batch_size=None,
    verbose='auto',
    sample_weight=None,
    steps=None,
    callbacks=None,
    max_queue_size=10,
    workers=1,
    use_multiprocessing=False,
    return_dict=False )
print("Test accuracy:", round(test_acc,2))

11/11 [==============================] - 0s 3ms/step - loss: 1.8558 - accuracy: 0.8441
Test accuracy: 0.84


    Here we are using squential modelling architecture from keras model and dense from keras layers. Sequential provides training and inference features on this model. There are many hidden layers of 64, 32 and 16 units. The model is then complied with rmsprop optimizer and set to a loss function of sparse_categorical_crossentropy. The model is then split into 80:20 % and it is trained with 80% of the data, 100 epoch with a batch size of 64. Finally the model is tested with remaining percentage of data and the accuracy was found to be 84%.